pkg/backends/datadog/datadog.go

```diff
@@ -191,17 +201,44 @@ func (d *Client) post(ctx context.Context, path, typeOfPost string, data interfa
 	}
 }
 
-func (d *Client) constructPost(ctx context.Context, authenticatedURL string, body []byte) func() error {
+func (d *Client) constructPost(ctx context.Context, authenticatedURL string, body []byte, compressPayload bool) func() error {
 	return func() error {
-		req, err := http.NewRequest("POST", authenticatedURL, bytes.NewReader(body))
+		headers := map[string]string{
+			"Content-Type": "application/json",
+			// Mimic dogstatsd code
+			"DD-Dogstatsd-Version": dogstatsdVersion,
+			"User-Agent":           dogstatsdUserAgent,
+		}
+		var reader io.Reader
+		if compressPayload {
+			// Use deflate (zlib, since DD requires the zlib headers)
+			var buf bytes.Buffer
+			compressor, err := zlib.NewWriterLevel(&buf, zlib.BestCompression)
+			if err != nil {
+				return fmt.Errorf("unable to create zlib writer: %v", err)
+			}
+			compressor.Write(body)
```

Comment:
This does not return an error?

Reply:
Also, this will be compressing the payload on each retry to send - see how the returned function is used. Compression code should probably be mored before the `return func()`.

Reply:
I considered that failures are uncommon but yes better to do it once. I've refactored.